docs/sources/developers/plugins/working-with-data-frames.md

```diff
@@ -0,0 +1,143 @@
++++
+title = "Working with data frames"
+type = "docs"
++++
+
+# Working with data frames
+
+The data frame is a columnar data structure, which allows efficient querying of large amounts of data. Data sources create data frames for panels to visualize. Since data frames are a central concept when developing plugins for Grafana, in this guide we'll look at some ways you can use them.
+
+The [DataFrame]({{< relref "../../packages_api/data/dataframe.md" >}}>) interface contains a `name` and an array of `fields` where each field contains the name, type, and the values for the field.
+
+Let's start with a simple example of a data frame representing a time series.
+
+Data frames representing time series contain at least a `time` field, and a `number` field. By convention, built-in plugins use `Time` and `Value` as field names for data frames containing time series data.
+
+```ts
+// Need to be of the same length.
+const timeValues = [1599471973065, 1599471975729];
+const numberValues = [12.3, 28.6];
+
+// Create data frame from values.
+const frame: DataFrame = {
+    name: "http_requests_total",
+    fields: [
+        { name: "Time", type: FieldType.time, values: timeValues },
+        { name: "Value", type: FieldType.number, values: numberValues }
+    ]
+    length: timeValues.length,
+}
```

Comment:
There's also the option to use `toDataFrame`here: https://github.com/grafana/grafana/blob/10141c5e6caeacaf817cd06ed1b1b8964791c410/packages/grafana-data/src/dataframe/processDataFrame.ts#L283

Not sure where this fits in though.

Reply:
How would you use toDataFrame here? How's it different from using ArrayDataFrame?

Reply:
I'm not an expert on this, but one difference is that it tries to discover the field types so you don't have to specify them yourself. Another difference is that it can handle legacy time series format conversions. And in your example there would be no need for the :DataFrame type. 

This is exactly what I find distracting/hard to reason about, which method to use when. So this doc is super important to align our api/thoughts :)
```suggestion
const frame = toDataFrame({
    name: "http_requests_total",
    fields: [
        { name: "Time", type: FieldType.time, values: timeValues },
        { name: "Value", type: FieldType.number, values: numberValues }
    ]
    length: timeValues.length,
})
```

Reply:
I believe that's what ArrayDataFrame does as well? In your suggestion, I don't see why the call to `toDataFrame` is needed, since I've already specified the fields and values?

Even though it's not needed, I added the `: DataFrame` annotation more as documentation to make it clear that it's fulfilling an interface. Too explicit?

Reply:
Oh I see now, `toDataFrame` returns either a MutableDataFrame or a ArrayDataFrame. In that case, I think we should prefer to use the actual classes rather than `toDataFrame` unless we're converting from legacy formats.

Reply:
toDataFrame does its best to convert the input into a DataFrame -- if checks all variations of what the input may be and returns it as a DataFrame

https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/dataframe/processDataFrame.ts#L283

MutableDataFrame will convert the values DTO (a simple array) to a Vector -- with `get()` functions etc

Reply:
> How's it different from using ArrayDataFrame?

ArrayDataFrame will take any array of objects and magic it into a DataFrame.

For ArrayDataFrame, you would start with:
```
const vals = [
{ time:123, value:1234, name:'abc' },
{ time:234, value:2345, name:'bcd' },
];
const frame = new ArrayDataFrame(vals);
```
That will look like a columnar data frame

Reply:
Yes, having looked at the code, `toDataFrame` looks like a smart factory that returns an implementation based on the input.

I think we should encourage people to understand and use MutableDataFrame and ArrayDataFrame directly rather than relying on the magic of `toDataFrame`. Using `toDataFrame` for legacy formats makes sense as you typically don't care. You just want to convert it.